canister/src/metrics.rs

```diff
@@ -18,6 +22,9 @@ pub struct Metrics {
 
     /// The total number of (valid) requests sent to `send_transaction`.
     pub send_transaction_count: u64,
+
+    // The number of instructions spent ingesting new Bitcoin blocks by type.
+    pub ingest_block_instructions_count: LabeledCounter,
```

Comment:
```suggestion
    // The number of instructions spent ingesting the most recent stable block.
    pub ingest_block_instructions_count: LabeledCounter,
```

Reply:
Are you sure about this comment, 'the most recent stable block' is just one block, while this counter will be the sum of instructions used for all ingested blocks?

Reply:
Oh I missed the part where we were summing all the instruction counts. I thought we were only keeping track of the most recent block ingestion, which is what would be more useful for our dashboards. I went ahead and made the following changes:

1. We're now only recording the most recent block ingestion (i.e. not summing)
2. Using a gauge instead of a counter (which is necessary since the value of the instruction counts can go up and down).
3. I updated the code to use the `ic_metrics_encoder` crate while we were at it - mostly simple changes.
4. I removed the proptest in `state.rs`, since now that we aren't summing the blocks anymore the test didn't contribute much.

Reply:
Regarding the first point: TBH, I am not convinced that recording the most recent block ingestion is better than summing, for example:

  1. one ingestion with I = (i1, i2, i3, ...) instructions used
  2. n ingestions with I = (i1, i2, i3, ...) instructions used
  
We will not see any difference on our Prometheus dashboards, and AFAIK it will not be possible to get the cumulative of instructions per type since we do not have any field to sum based on it.

Regarding the last point, that test was intended to verify that metrics are correctly updated after executing `ingest_block`, without that test if we delete the places on which we change the value of the metric, all tests will pass, but we do not have the intended behavior.

Reply:
Regarding 3rd point, thanks for changing that.

Reply:
@dragoljub-duric As discussed offline, I kept the gauge and re-added the test in `state.rs`. I modified it a bit so that it's testing specifically the two cases where we update the stats.